File: Sean的leetcode/Leetcode/Palindrome Permutation.cpp
/*
Time Complexity:O(n)
*/
class Solution {
public:
    bool canPermutePalindrome(string s) {
        vector<int> cnt(256,0);
        int n = s.size();
        if (n == 0) return true;
        int oddcnt = 0;
        for (char c: s) {
            cnt[c]++;
            if (cnt[c] & 1) {
                oddcnt++;
            }
            else {
                oddcnt--;
            }
        }
        
        return oddcnt <= 1;
    }
};
//Sean's solution
class Solution {
public:
    bool canPermutePalindrome(string s) {
        unordered_map<char, int> c_map;
        int odd = 0;
        for (auto c : s)
        {
            c_map[c]++;
            if (c_map[c] & 1)
                odd++;
            else
                odd--;
        }
        return odd <= 1;
    }
};

File: Sean的leetcode/python/Palindrome Permutation.py
###
# Hash Table
# Time Complexity: O(n) 
# Space Complexity: O(1)
###
class Solution(object):
    def canPermutePalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        from collections import Counter
        if not s:
            return False
        odd = sum(n & 1 for n in Counter(s).values())
        return odd <= 1
        
class Solution(object):
    def canPermutePalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        from collections import defaultdict
        if not s:
            return False
        d = defaultdict(int)
        odd = 0
        for c in s:
            d[c] += 1
            if d[c] & 1:
                odd += 1
            else:
                odd -= 1
        return odd <= 1
        
        